fuel/datasets/hdf5.py

```diff
@@ -316,11 +316,14 @@ def _out_of_memory_get_data(self, state=None, request=None):
             if isinstance(request, slice):
                 request = slice(request.start + subset.start,
                                 request.stop + subset.start, request.step)
+                data = state[source_name][request]
             elif isinstance(request, list):
                 request = [index + subset.start for index in request]
+                sorted_request = numpy.sort(request)
+                sorted_data = state[source_name][sorted_request, ...]
+                data = sorted_data[numpy.argsort(numpy.argsort(request))]
```

Comment:
What if you pre-allocated `data` and did

`data[indices] = state[source_name][request[indices], ...]`